net: Use actual memory size in receive buffer accounting

laanwj opened the pull request:
Add a method `CNetMessage::GetMemoryUsage` and use this for accounting of the size of the process receive queue instead of the raw message size (like we already do for the send buffer and `CSerializedNetMsg`).

This ensures that allocation and deserialization overhead is better taken into account. 

On average, this counts about ~100 extra bytes per packet on x86_64:
```
2024-10-27T09:50:12Z [net] 24 bytes -> 112 bytes
2024-10-27T10:36:37Z [net] 61 bytes -> 176 bytes
2024-10-27T10:36:38Z [net] 1285 bytes -> 1392 bytes
2024-10-27T09:50:21Z [net] 43057 bytes -> 43168 bytes
```

laanwj committed:
memusage: Allow counting usage of vectors with different allocators

laanwj committed:
streams: add DataStream::GetMemoryUsage

laanwj committed:
net: Use actual memory size in receive buffer accounting

Add a method CNetMessage::GetMemoryUsage and use this for accounting of
the size of the process receive queue instead of the raw message size.

This ensures that allocation and deserialization overhead is taken into
account.

theStack reviewed (approved):
ACK ad5c012157c5f261503022cfa22d7124bfda5765

> On average, this counts about ~100 extra bytes per packet on x86_64:
> 
> ```
> 2024-10-27T09:50:12Z [net] 24 bytes -> 112 bytes
> 2024-10-27T10:36:37Z [net] 61 bytes -> 176 bytes
> 2024-10-27T10:36:38Z [net] 1285 bytes -> 1392 bytes
> 2024-10-27T09:50:21Z [net] 43057 bytes -> 43168 bytes
> ```

Seeing very similar numbers here w.r.t. to the extra byte count (tested on OpenBSD 7.6, x86_64):
```
2024-11-01T01:25:18Z msg type: inv, raw size: 58, mem usage: 168
2024-11-01T01:25:18Z msg type: inv, raw size: 130, mem usage: 232
2024-11-01T01:25:18Z msg type: inv, raw size: 166, mem usage: 280
2024-11-01T01:25:19Z msg type: tx, raw size: 742, mem usage: 856
2024-11-01T01:25:19Z msg type: tx, raw size: 3010, mem usage: 3112
2024-11-01T01:25:19Z msg type: tx, raw size: 17758, mem usage: 17864
```
<details>
<summary>Patch</summary>

```diff
diff --git a/src/net.cpp b/src/net.cpp
index 49edceead4..2c5725e299 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -3779,6 +3779,7 @@ void CNode::MarkReceivedMsgsForProcessing()
         // vRecvMsg contains only completed CNetMessage
         // the single possible partially deserialized message are held by TransportDeserializer
         nSizeAdded += msg.GetMemoryUsage();
+        LogPrintf("msg type: %s, raw size: %d, mem usage: %d\n", msg.m_type, msg.m_raw_message_size, msg.GetMemoryUsage());
     }
 
     LOCK(m_msg_process_queue_mutex);
```
</details>

yuvicc reviewed (approved):
Tested ACK ad5c012157c5f261503022cfa22d7124bfda5765

> Add a method `CNetMessage::GetMemoryUsage` and use this for accounting of the size of the process receive queue instead of the raw message size (like we already do for the send buffer and `CSerializedNetMsg`).
> 
> This ensures that allocation and deserialization overhead is better taken into account.
> 
> On average, this counts about ~100 extra bytes per packet on x86_64:
> 
> ```
> 2024-10-27T09:50:12Z [net] 24 bytes -> 112 bytes
> 2024-10-27T10:36:37Z [net] 61 bytes -> 176 bytes
> 2024-10-27T10:36:38Z [net] 1285 bytes -> 1392 bytes
> 2024-10-27T09:50:21Z [net] 43057 bytes -> 43168 bytes
> ```

Tested on Ubuntu 22.04.4 LTS x86_64 (extra bytes ~100)
```
2024-11-02T16:09:18Z msg: pong, raw size: 29, mem usage: 144
2024-11-02T16:09:18Z msg: headers, raw size: 103, mem usage: 224
2024-11-02T16:09:18Z msg: sendheaders, raw size: 33, mem usage: 112
2024-11-02T16:09:18Z msg: cmpctblock, raw size: 23280, mem usage: 23392
2024-11-02T16:09:18Z msg: pong, raw size: 32, mem usage: 144
```

laanwj commented on src/net.cpp:
```diff
@@ -127,6 +127,12 @@ size_t CSerializedNetMsg::GetMemoryUsage() const noexcept
 
+size_t CNetMessage::GetMemoryUsage() const noexcept
+{
+    // Don't count the dynamic memory used for the m_type string (see CSerializedNetMsg::GetMemoryUsage),
```
The comment in the referred function explains it.
```
    // Don't count the dynamic memory used for the m_type string, by assuming it fits in the
    // "small string" optimization area (which stores data inside the object itself, up to some
    // size; 15 bytes in modern libstdc++).
```
 It has to do with the internal implementation of std::string, where it has a fixed-size buffer that it will use for short strings (such as the message types) before allocating memory on the heap.

laanwj replied:
i do agree that over-counting is probably better than under-counting here, but i just took this over from `CSerializedNetMsg` as i assumed this was the right thing to do. If we change it it needs to be changed in both places.

laanwj replied:
For g++ 13.2 on x86_64 it checks out, this returns 16:
```c++
#include <string>
#include <iostream>

bool data_is_internal(const std::string &s)
{
    return s.data() >= reinterpret_cast<const char*>(&s) && s.data() + s.size() - sizeof(s) <= reinterpret_cast<const char*>(&s);
}

int main()
{
    std::string s;
    while (data_is_internal(s)) {
        s += 'a';
    }
    std::cout << s.size() << std::endl;
    return 0;
}
```
But yes it's not possible to guarantee across c++ compilers and platforms so adding the check to `DynamicUsage` makes sense imo.

l0rinc replied:
Alternatively, maybe a test checking this exact scenario could help:
```C++
BOOST_AUTO_TEST_CASE(sso_size_test)
{
    CSerializedNetMsg empty_msg;

    CSerializedNetMsg msg_with_type;
    msg_with_type.m_type = std::string(CMessageHeader::COMMAND_SIZE, 'X');

    BOOST_CHECK_EQUAL(sizeof(msg_with_type), sizeof(empty_msg));
}
```

l0rinc replied:
Thanks for the hints, @sipa!

According to https://en.cppreference.com/w/cpp/utility/compare/compare_three_way, this also seems safe to use:
```c++
return (msg_start <=> s.data()) <= 0 && (s.data() <=> msg_end) <= 0;
```
or
```c++
return (msg_start <=> s.data()) != std::strong_ordering::greater
      && (s.data() <=> msg_end) != std::strong_ordering::greater;
```

l0rinc replied:
> i'm kind of hestitant to add a test like that because i'm not sure we want to require this behavior from the C++ library.

I understood the memory calculations will be incorrect if this would fail - is that not the case?

laanwj replied:
> I understood the memory calculations will be incorrect if this would fail - is that not the case?

Yes, unless we add the logic to `memusage.h`. i guess i've always seen memusage more as a "best effort estimate" than something that needs to be correct to the byte, because it's so hard to guarantee across implementations, we don't really want to extend our scope to C++ library implementation specifics. But i dunno.

BTW: It looks like currently. memusage.h doesn't handle `std::string` (nor `std::basic_string` at all. This would be new.

laanwj replied:
Done

laanwj commented on src/memusage.h:
```diff
@@ -84,8 +84,8 @@ struct stl_shared_counter
+template<typename X, typename Y>
+static inline size_t DynamicUsage(const std::vector<X, Y>& v)
 {
     return MallocUsage(v.capacity() * sizeof(X));
```
Right, it can,  when `reserve()` is called with a larger amount, or when the vector is shortened. But it does represent the actually-allocated size which seems what we want here.

laanwj commented on src/memusage.h:
```diff
@@ -84,8 +84,8 @@ struct stl_shared_counter
+template<typename X, typename Y>
+static inline size_t DynamicUsage(const std::vector<X, Y>& v)
 {
     return MallocUsage(v.capacity() * sizeof(X));
```
Correct, it's computes the dynamic usage only.

laanwj commented on src/net.h:
```diff
@@ -245,6 +245,9 @@ class CNetMessage
     CNetMessage& operator=(CNetMessage&&) = default;
     CNetMessage& operator=(const CNetMessage&) = delete;
+
+    /** Compute total memory usage of this object (own memory + any dynamic memory). */
```
It's the total memory, making the assumption about the underlying implementation of std::string. Could add that here too, sure.

laanwj commented on src/net.cpp:
```diff
@@ -3772,7 +3778,7 @@ void CNode::MarkReceivedMsgsForProcessing()
     for (const auto& msg : vRecvMsg) {
         // vRecvMsg contains only completed CNetMessage
         // the single possible partially deserialized message are held by TransportDeserializer
-        nSizeAdded += msg.m_raw_message_size;
```
It's still useful for one purpose: to count bytes going over the network for bandwidth monitoring. This icnludes the statistics per message type.

l0rinc replied:
Got it, thanks for clarifying!

laanwj commented on src/memusage.h:
```diff
@@ -84,8 +84,8 @@ struct stl_shared_counter
-template<typename X>
-static inline size_t DynamicUsage(const std::vector<X>& v)
+template<typename X, typename Y>
+static inline size_t DynamicUsage(const std::vector<X, Y>& v)
```
https://en.cppreference.com/w/cpp/container/vector uses
```
template<
    class T,
    class Allocator = [std::allocator](http://en.cppreference.com/w/cpp/memory/allocator)<T>
> class vector
```
so let's go with those names

laanwj committed:
memusage: Add DynamicUsage for std::string

Add DynamicUsage(std::string) which Returns the dynamic allocation of a std::string,
or 0 if none (in case of small string optimization).

laanwj committed:
net: Use DynamicUsage(m_type) in CSerializedNetMsg::GetMemoryUsage

Now that memusage correctly computes the dynamic size of a string, there
is no need for special handling here.

l0rinc commented on src/memusage.h:
```diff
@@ -90,6 +90,18 @@ static inline size_t DynamicUsage(const std::vector<T, Allocator>& v)
+    const char* s_ptr = reinterpret_cast<const char*>(&s);
+    // Don't count the dynamic memory used for string, if it resides in the
+    // "small string" optimization area (which stores data inside the object itself, up to some
+    // size; 15 bytes in modern libstdc++).
```
note: with clang I measured ~24: https://tc-imba.github.io/posts/cpp-sso

laanwj replied:
interesting!
that also suggests that on 32-bit platforms, the size would still be 15 bytes for g++ (because it's the minimum bound), but 10 (3*4 minus 2 for size byte and zero byte) for clang
so it would trigger an allocation for message types of 11/12 chars there 😄

laanwj replied:
i don't think there's anything to do here; the code is agnostic to the size and implementation of the sso area; the comment gives an example for one C++ library, to illustrate the point
adding more specifics adds more information that could go stale 😄

l0rinc commented on src/memusage.h:
```diff
@@ -90,6 +91,18 @@ static inline size_t DynamicUsage(const std::vector<T, Allocator>& v)
+    // Don't count the dynamic memory used for string, if it resides in the
+    // "small string" optimization area (which stores data inside the object itself, up to some
+    // size; 15 bytes in modern libstdc++).
+    if (!std::less{}(s.data(), s_ptr) && !std::greater{}(s.data() + s.size(), s_ptr + sizeof(s))) {
```
Since I think these are related pointers (unlike https://github.com/bitcoin/bitcoin/pull/31164#discussion_r1827884684), can we use simple pointer comparisons here? Or do I misunderstand how this works?
```C++
if (s.data() >= s_ptr && s.data() + s.size() <= s_ptr + sizeof(s)) {
```

laanwj replied:
as i understand it, in the case that sso is not used, these are not related pointers

l0rinc replied:
Thanks for clarifying.

laanwj replied:
fwiw the underlying reason for this pointer absurdity is to support platforms with segmented memory (such as GPUs), where pointers to different types of allocations might refer to different address spaces and won't be comparable
probably, `std::greater` adds some address space prefix to give them a defined order
this is a super edge case and unlikely to come up for any platform people compile bitcoind for, but better to be safe i guess...

l0rinc commented on src/memusage.h:
```diff
@@ -90,6 +91,18 @@ static inline size_t DynamicUsage(const std::vector<T, Allocator>& v)
     return MallocUsage(v.capacity() * sizeof(T));
 }
 
+static inline size_t DynamicUsage(const std::string& s)
```
Could you please add a test in this commit that demonstrates the intended behavior? Or would that tie the implementation down too much?

laanwj replied:
i guess we could do some basic generalizable test like "the memory usage increases or stays equal when the string becomes longer"
but i'm not sure how useful that is, and there are currently no memusage tests to add it to

laanwj replied:
i've tested it manually (gcc, x86_64) with the following patch:
<details>
  <summary>patch</summary>

```patch
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 8a4c1746161c30c04dee1f41a811740d0038df0b..f8c688dd3617123548af13611519adba59eef775 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -23,6 +23,7 @@
 #include <util/strencodings.h>
 #include <util/time.h>
 #include <util/translation.h>
+#include <memusage.h>

 #include <algorithm>
 #include <chrono>
@@ -1281,6 +1282,13 @@ static int CommandLineRPC(int argc, char *argv[])

 MAIN_FUNCTION
 {
+    std::string s;
+    std::cout << "length (b)" << "," << "dynsize (b)" << std::endl;
+    for (int x=0; x<=200; ++x) {
+        std::cout << x << "," << memusage::DynamicUsage(s) << std::endl;
+        s += 'a';
+    }
+    exit(0);
 #ifdef WIN32
     common::WinCmdLineArgs winArgs;
     std::tie(argc, argv) = winArgs.get();
```

</details>

as expected, for the first 15 bytes it stays flat at 0, then it allocates and the first allocation is 48 bytes, and it increases stepwise as the size of the string grows:
![image](https://github.com/user-attachments/assets/7ba2adec-b0be-4e31-bafe-a8b9386af222)

l0rinc reviewed (commented):
Concept ACK

laanwj commented on src/memusage.h:
```diff
@@ -84,8 +84,8 @@ struct stl_shared_counter
+template<typename X, typename Y>
+static inline size_t DynamicUsage(const std::vector<X, Y>& v)
 {
     return MallocUsage(v.capacity() * sizeof(X));
```
yes, explaining it more would make sense!

but i think it's out of scope for this PR

laanwj commented on src/memusage.h:
```diff
@@ -84,8 +84,8 @@ struct stl_shared_counter
 
-template<typename X>
-static inline size_t DynamicUsage(const std::vector<X>& v)
+template<typename X, typename Y>
```
right-it can be used for nested structures, but the call site has to do their own iteration and add it up, as this only covers the outermost layer

adding the `std::is_trivial_v` check is an interesting idea but would prevent it from being used there.

l0rinc replied:
In a next PR we could add another DynamicUsage which accepts a vector of vectors and does the iteration instead of the call sites

l0rinc commented:
ACK d22a234ed270286b483aec2db1e2f716b9756231

yuvicc reviewed (approved):
ACK d22a234ed270286b483aec2db1e2f716b9756231

danielabrozzoni reviewed (approved):
Light ACK d22a234ed270286b483aec2db1e2f716b9756231 - code looks good to me, but I'm not very familiar with C++ memory management specifics

achow101 commented:
ACK d22a234ed270286b483aec2db1e2f716b9756231

theStack reviewed (commented):
post-merge ACK d22a234ed270286b483aec2db1e2f716b9756231
